accepted/2024/trusted-publishers-oidc-for-nuget-push.technical.md

```diff
@@ -0,0 +1,189 @@
+# Trusted Publishers, using OpenID Connect for NuGet push (technical)
+
+- Author: Joel Verhagen ([@joelverhagen](https://github.com/joelverhagen) on GitHub)
+- Issue: [NuGet/NuGetGallery#9332](https://github.com/NuGet/NuGetGallery/issues/9332)
+
+This is the technical description of the experience described in [Trusted Publishers, using OpenID Connect for NuGet
+push](trusted-publishers-oidc-for-nuget-push.md). The content is split to clearly separate the functional/UX description
+(the other doc) from the technical description (this doc). This document acts as a supporting document to the other so
+that other one should be read first.
+
+Much of the technical explanation is described in [Trusted Publishers for All Package
+Repositories](https://repos.openssf.org/trusted-publishers-for-all-package-repositories). I will expand on certain
+technical details which are particularly interesting or specific to NuGet.
+
+## Validation of the GitHub OIDC token
+
+The most critical step in this design is understanding how NuGet.org will validate that an incoming OIDC token is
+acceptable to trade for a short-lived NuGet API key.
+
+The following checks are made:
+
+- The token is issued by a known Trusted Publisher (only GitHub Actions at this time)
+- The token is validated per all JWT rules, such as but not limited to:
+  - valid signature via JWKS
+  - validate duration (`nbf` and `exp` claims)
+  - only used once, via `jti`
+  - valid `aud` claim, being `nuget`
+
+The
+[Microsoft.IdentityModel.Protocols.OpenIdConnect](https://www.nuget.org/packages/Microsoft.IdentityModel.Protocols.OpenIdConnect)
+package can help us properly validate JWTs in our NuGet/NuGetGallery ASP.NET application.
+
+In addition to the general JWT checks, specific checks are made for each Trusted Publisher. For GitHub Actions, the
+following checks will be made:
+
+- `sub` claim matches `{repo owner}/{repo name}:.*` (case insensitive)
+  - The suffix of the `sub` is implied by the other checks
+- `repository_owner` claim matches the `{repo owner}` name (case insensitive)
+- `repository_owner_id` claim matches the numeric owner ID recorded at the time of the trust policy creation
```

Comment:
This numeric owner ID is forge-assigned (here GitHub-assigned)?

Reply:
Yes, GitHub assigned.

I should note that other forges may not have a numeric owner ID. We will need to assess resurrection attacks or any other impersonation risks on a case by case basis.